Tracing support for Lambda extension

DarcyRaynerDD opened the pull request:
### What does this PR do?

This PR makes some modifications to the Trace Agent to allow traces to be flushed synchronously. This synchronous flushing mode is used by the [Lambda Extension](https://github.com/DataDog/devops/wiki/Lambda-Extension) to time when traces are submitted to the backend in order to minimize latency. This is described in this [draft RFC](https://github.com/DataDog/architecture/pull/762).

I created a generic interface for TraceWriter and StatsWriter in TraceAgent, and created a  StatsSyncWriter and TraceSyncWriter implementation, which supports a new Flush method. Calling Flush will wait for any inflight requests to be processed and submitted before continuing.

The agent will then swap to using the Sync writers when either using the Extension or 'apm_config.sync_flushing' is set to true. Otherwise it will rely on the existing writers.

### Motivation

Support for tracing in lambda using the AWS Extension will make it easier to trace AWS Lambda function.

### Additional Notes

Is StatsSyncWriter necessary if we are using backend stats calculations? Could this be a no-op writer instead?

### Describe your test plan

I added some unit tests for the parts that affect the trace agent. For the parts that only affect the Extension I've manually tested:
- Submitting traces via the extension
- Trace obfuscations, and using datadog.yaml with APM settings.

DarcyRaynerDD committed:
[serverless] Add SyncAgent support for serverless extension

remeh committed:
Add the Serverless Datadog Agent implementation.

See https://docs.datadoghq.com/serverless/datadog_lambda_library/extension/

remeh committed:
Merge remote-tracking branch 'origin/master' into remeh/serverless

remeh committed:
Fix 3rd party license list since new pkg are imported from the AWS SDK.

remeh committed:
forwarder: adapt the SyncDefaultForwarder to recent changes in the Forwarder interface.

remeh committed:
aggregator: fix unit tests with Flush now being a public function.

remeh committed:
serverless: linter compliance.

remeh committed:
serverless: errcheck compliance.

remeh committed:
serverless: unused code linter.

remeh committed:
serverless: linter compliance.

DarcyRaynerDD committed:
[serverless] implement PR feedback

DarcyRaynerDD committed:
[serverless] Refactor Agent to support sync flush mode

DarcyRaynerDD committed:
[serverless] fix compiler errors

DarcyRaynerDD committed:
[trace][serverless] add tests to TraceSyncWriter

DarcyRaynerDD committed:
[trace][serverless] Add StatsSyncWriter tests

DarcyRaynerDD committed:
Merge branch 'master' into darcy.rayner/trace-lambda-extension

DarcyRaynerDD committed:
Merge branch 'master' into darcy.rayner/trace-lambda-extension

DarcyRaynerDD committed:
Fix merge issues

DarcyRaynerDD committed:
[serverless] remove unnecessary diffs

DarcyRaynerDD committed:
[serverless] Use config to define extension tracing

DarcyRaynerDD committed:
[serverless] move constant location

gbbr commented on pkg/trace/writer/stats_sync.go:
```diff
@@ -0,0 +1,208 @@
+const maxPendingStatsPayloads = 10
+
+// StatsSyncWriter ingests stats buckets and flushes them to the API.
+type StatsSyncWriter struct {
```
There seems to be a lot of duplication here. Can't we have the original trace/stats writers and have them simply take a flag which alters their behaviour? I mean 90% of this code is the same as the original one. That way we could also get rid of the interfaces.

DarcyRaynerDD replied:
@gbbr I've gone through and merged the logic from the two files. I think you're right, it is neater.

remeh commented on pkg/trace/writer/trace_sync.go:
```diff
@@ -0,0 +1,233 @@
+		maxmem := cfg.MaxMemory / 2
+		if maxmem == 0 {
+			// or 500MB if unbound
+			maxmem = 500 * 1024 * 1024
```
Ditto

remeh reviewed (commented):
Great work! Integration within the Serverless Agent code LGTM, as for the code duplication, I agree that using an interface is the more idiomatic way of doing that, however in this case it indeed includes a lot of code duplication. If we can't share that code between these two interfaces, as Gabriel mentioned, that may be a better trade-off to just use a flag in the original code?

DarcyRaynerDD committed:
[serveless] refactor sync trace writer into trace writer

DarcyRaynerDD committed:
[serverless] remove stats sync writer

DarcyRaynerDD committed:
[serverless] reduce unnecessary pr diffs

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,13 @@ func (a *Agent) Run() {
 
+// Flush traces sychronously. This method only works when the agent is configured in synchronous flushing
+// mode.
+func (a *Agent) Flush() {
```
Should we call it `FlushSync`?

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -78,6 +78,7 @@ type AgentConfig struct {
 	MaxRequestBytes int64 // specifies the maximum allowed request size for incoming trace payloads
 
 	// Writers
+	SynchronousFlushing     bool
```
This warrants some prose. Later on it will be hard to figure out what this is.

gbbr commented on pkg/trace/config/apply.go:
```diff
@@ -280,6 +280,9 @@ func (c *AgentConfig) applyDatadogConfig() error {
 		c.ConnectionResetInterval = getDuration(config.Datadog.GetInt("apm_config.connection_reset_interval"))
 	}
+	if config.Datadog.IsSet("apm_config.sync_flushing") {
+		c.SynchronousFlushing = config.Datadog.GetBool("apm_config.sync_flushing")
```
You should add this as know to this function: https://github.com/DataDog/datadog-agent/blob/90d4d955067ca0e43764aa339112f007bb4836b3/pkg/config/apm.go. And also add an env. var for it?

It might also have to be in the config template along with proper documentation: https://github.com/DataDog/datadog-agent/blob/90d4d955067ca0e43764aa339112f007bb4836b3/pkg/config/config_template.yaml. But I'll let you decide.

DarcyRaynerDD replied:
I think I'll add it to setKnown, but I don't think I'll add it to the template because user's of the regular trace agent shouldn't toggle it.

gbbr replied:
Goti it.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -36,25 +37,32 @@ const (
+	wg               sync.WaitGroup
+	stats            *info.StatsWriterInfo
+	payloads         []*stats.Payload // payloads buffered for sync mode
+	flushSyncEnabled bool
```
Can we use `sync` or `syncMode`? It's shorter :)

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,31 @@ func (w *StatsWriter) Run() {
+	w.wg.Wait()
+
+	// Wait for all the senders to finish
+	wg := sync.WaitGroup{}
```
```suggestion
	var wg sync.WaitGroup
```

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,31 @@ func (w *StatsWriter) Run() {
+	w.flushChan <- struct{}{}
+	w.wg.Wait()
+
+	// Wait for all the senders to finish
```
I don't think this comment is needed. We're not waiting here, we're waiting further down.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,31 @@ func (w *StatsWriter) Run() {
+	for _, sender := range w.senders {
+		s := sender
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			s.waitForInflight()
+		}()
```
I think the common pattern here is:
```suggestion
		go func(s *sender) {
			defer wg.Done()
			s.waitForInflight()
		}(sender)
```
I'm not even sure if declaring the local variable works. It may, but the pattern is generally the above one.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,31 @@ func (w *StatsWriter) Run() {
+		}()
+	}
+	wg.Wait()
+	return nil
```
I think we can remove the return value. This never returns anything else than `nil`.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,31 @@ func (w *StatsWriter) Run() {
+
+	defer w.report()
+	w.wg.Add(1)
+	w.flushChan <- struct{}{}
```
I don't really get this part. Why is this here?

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,31 @@ func (w *StatsWriter) Run() {
+// FlushSync is a no-op for TraceWriter
+func (w *StatsWriter) FlushSync() error {
+	if !w.flushSyncEnabled {
+		log.Error("SyncFlush called on StatsWriter, which is a no-op")
```
If it's a no-op, let's return early.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -136,6 +172,13 @@ func (w *StatsWriter) SendPayload(p *stats.Payload) {
+	for _, p := range w.payloads {
+		w.SendPayload(p)
+	}
+	w.payloads = []*stats.Payload{}
```
Maybe it's better to do this instead of re-allocating a new slice
```suggestion
	w.payloads = w.payloads[:0]
```

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +144,50 @@ func (w *TraceWriter) Run() {
+			w.addSpans(pkg)
+		case <-w.flushChan:
+			// Make sure all pending spans have been buffered
+			w.processPendingSpans()
```
Let's call this method `w.drainAndFlush`

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +144,50 @@ func (w *TraceWriter) Run() {
+	w.wg.Wait()
+
+	// Wait for all the senders to finish
+	wg := sync.WaitGroup{}
```
Same comments as above on this function

gbbr reviewed (commented):
Left some initial thoughts. And I agree. This *is* much better. Thanks for taking the time to do it.

I have some concerns: 
- About the naming "synchronous flushing". I'm not sure if it's the right terminology. What actually happens is that there is no flushing at all until the user triggers it. In this case, the question comes, why not modify just the sender structure? Have everything work exactly as it did until today, but only flush on command. That way you only need to change the `sender` and not both the trace and stats writer. I might be missing something so apologies if I do.

DarcyRaynerDD committed:
[serverless] address pr feedback

DarcyRaynerDD committed:
[serverless] address more pr feedback

DarcyRaynerDD committed:
[serverless] fix compile error

gbbr commented on pkg/config/apm.go:
```diff
@@ -41,6 +41,7 @@ func setupAPM(config Config) {
 	config.SetKnown("apm_config.log_throttling")
 	config.SetKnown("apm_config.bucket_size_seconds")
 	config.SetKnown("apm_config.watchdog_check_delay")
+	config.SetKnown("apm_config.sync_flushing")
```
Are we going to want to allow this to be set by means of an env. var. ? If so, we might want to bind it here in the same function to `DD_APM_SYNC_FLUSHING`, or `DD_APM_LAMBDA`, whichever suits the use case more.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -36,25 +37,31 @@ const (
+	flushChan chan struct{}
+	wg        sync.WaitGroup
+	stats     *info.StatsWriterInfo
+	payloads  []*stats.Payload // payloads buffered for sync mode
+	syncMode  bool
```
A small nit for making more sense of this. Do you mind if we add an empty line between these two and the rest? They are like a separate category exclusively for sync. Also reversing the order might make sense. I mean something like:
```suggestion

	syncMode  bool             // sync mode enabled -- writer only flushes when FlushSync is called.
	payloads  []*stats.Payload // payloads buffered for sync mode
```

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +108,30 @@ func (w *StatsWriter) Run() {
+// FlushSync is a no-op for TraceWriter
+func (w *StatsWriter) FlushSync() error {
+	if !w.syncMode {
+		return log.Error("SyncFlush called on StatsWriter, which is a no-op")
```
```suggestion
		return errors.New("Synchronous flushing is disabled. Calling FlushSync will have no effect.")
```
I think returning the error alone is enough. The logging should happen in `(*Agent).FlushSync`

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +144,49 @@ func (w *TraceWriter) Run() {
+		case pkg := <-w.In:
+			w.addSpans(pkg)
+		case <-w.flushChan:
+			// Make sure all pending spans have been buffered
```
I think we can leave this comment out. The method name on the next line is very selfexplanatory.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +144,49 @@ func (w *TraceWriter) Run() {
+			s.waitForInflight()
+		}(s)
+	}
+	// Wait for all the senders to finish inflight requests
```
Same here I think the pattern is very obvious, no need to comment.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +144,49 @@ func (w *TraceWriter) Run() {
+	}
+	defer w.report()
+
+	w.wg.Add(1) // Add will be completed in sync loop
+	w.flushChan <- struct{}{}
+	// Wait for all payloads to be submitted to the sender
+	w.wg.Wait()
```
I really don't like this approach. Seems like the wrong use for the concept. There is no "group" here :) How about we make `flushChan` of the type `chan chan struct{}` and use it as such:
```go
notify := make(chan struct{}, 1) // we don't need to block the select statement
w.flushChan <- notify
<-notify
```
Should be a simply synchronisation mechanism, and we can get rid of the struct field. In the select statement, you would pick up the channel and make sure to notify back when done.

DarcyRaynerDD replied:
Sounds good to me

gbbr reviewed (commented):
Thanks Darcy! I think we're getting closer :)

DarcyRaynerDD committed:
[serverless] implement pr feedback

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +145,49 @@ func (w *TraceWriter) Run() {
+		select {
+		case pkg := <-w.In:
+			w.addSpans(pkg)
+		case notify := <-w.flushChan:
```
Seems like the only difference between the two methods (`runSync` and `runAsync`) is the `flushChan` case, so we can maybe stick to just one method `Run`. Just like we had before, and just like you did in the stats writer? There would be an extra case for `flushChan` but that wouldn't be triggered if sync mode is off. WDYT? Or did I miss something?

DarcyRaynerDD replied:
The runAsync also uses a ticker for flushing, which we don't want in the runSync loop. That's the only other difference, but it meant we couldn't use the same select statement for both.

gbbr replied:
Ah yes. Doh. Sorry I missed that. Resolving.

gbbr reviewed (commented):
A few more nits.

DarcyRaynerDD committed:
[serverless] add _dd.origin:lambda to extension traces

DarcyRaynerDD committed:
Merge branch 'master' into darcy.rayner/trace-lambda-extension

DarcyRaynerDD committed:
[serverless] more PR feedback

DarcyRaynerDD commented:
@gbbr Can I get another review. I added a metadata tag to the trace payload, called `__dd.origin:lambda`, which is used by the trace intake and the frontend to detect lambda traces to apply special cases. I added an `AdditionalMetadata` field to the agent config, and made the agent add that metadata to each span. Please let me know if there is a cleaner way to do this.

DarcyRaynerDD committed:
[serverless] fix linting error

DarcyRaynerDD committed:
[serverless] remove unused constant

DarcyRaynerDD committed:
Merge branch 'master' into darcy.rayner/trace-lambda-extension

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,21 @@ func (a *Agent) Run() {
 }
 
+// FlushSync flushes traces sychronously. This method only works when the agent is configured in synchronous flushing
+// mode.
```
```suggestion
// mode via the apm_config.sync_flush option.
```

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -193,6 +208,9 @@ func (a *Agent) Process(p *api.Payload, sublayerCalculator *stats.SublayerCalcul
 		// Extra sanitization steps of the trace.
 		for _, span := range t {
+			for k, v := range a.conf.AdditionalMetadata {
+				traceutil.SetMeta(span, k, v)
```
Do you want this on every span of the trace? Why not only the root? It's totally up to you but just making sure this is what you want :)

DarcyRaynerDD replied:
Yeah, we need to apply __dd.origin:lambda to every span in the trace. It enables special processing/displaying on the trace intake side.

gbbr commented on pkg/trace/config/config.go:
```diff
@@ -107,6 +108,9 @@ type AgentConfig struct {
 	// It maps tag keys to a set of replacements. Only supported in A6.
 	ReplaceTags []*ReplaceRule
 
+	// AdditonalMetadata list metadata that will be added to all spans
```
```suggestion
	// GlobalTags lists tags that will be attached to all spans.
```
Can we please call it GlobalTags? This is simply because we are already using this terminology for the same functionality in tracers. It will be easier to follow.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -36,25 +38,31 @@ const (
+	flushChan chan chan struct{}
+	stats     *info.StatsWriterInfo
+
+	syncMode bool
```
```suggestion
	// syncMode reports whether the writer should flush on its own or only when FlushSync is called
	syncMode bool
```

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +109,29 @@ func (w *StatsWriter) Run() {
+// FlushSync is a no-op for TraceWriter
+func (w *StatsWriter) FlushSync() error {
+	if !w.syncMode {
+		return errors.New("SyncFlush called on StatsWriter, which is a no-op")
```
```suggestion
		return errors.New("not flushing; sync mode not enabled")
```

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,21 @@ func (a *Agent) Run() {
+func (a *Agent) FlushSync() {
+	err := a.StatsWriter.FlushSync()
+	if err != nil {
+		log.Errorf("Error flushing trace agent: %s", err.Error())
```
```suggestion
		log.Errorf("Error flushing stats: %s", err.Error())
```

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,21 @@ func (a *Agent) Run() {
+	}
+	err = a.TraceWriter.FlushSync()
+	if err != nil {
+		log.Errorf("Error flushing trace agent: %s", err.Error())
```
```suggestion
		log.Errorf("Error flushing traces: %s", err.Error())
```

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,21 @@ func (a *Agent) Run() {
+		log.Errorf("Error flushing trace agent: %s", err.Error())
+		return
+	}
+	err = a.TraceWriter.FlushSync()
+	if err != nil {
```
```suggestion
	if err := a.TraceWriter.FlushSync(); err != nil {
```

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,21 @@ func (a *Agent) Run() {
+// FlushSync flushes traces sychronously. This method only works when the agent is configured in synchronous flushing
+// mode.
+func (a *Agent) FlushSync() {
+	err := a.StatsWriter.FlushSync()
+	if err != nil {
```
```suggestion
	if err := a.StatsWriter.FlushSync(); err != nil {
```

gbbr commented on pkg/trace/agent/agent.go:
```diff
@@ -112,6 +112,21 @@ func (a *Agent) Run() {
 
+// FlushSync flushes traces sychronously. This method only works when the agent is configured in synchronous flushing
+// mode.
+func (a *Agent) FlushSync() {
```
Let's just return immediately if it's disabled.
```suggestion
func (a *Agent) FlushSync() {
    if a.conf.SyncFlushing {
        log.Error("(*Agent).FlushSync called without apm_conf.sync_flushing enabled. No data was sent to Datadog.")
        return
    }
```
I'm wondering if we should crash completely because of this, as a critical error. It would seem like one. If you think so, simply change to `log.Critical`

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +109,29 @@ func (w *StatsWriter) Run() {
+	w.flushChan <- notify
+	<-notify
+
+	var wg sync.WaitGroup
+	for _, s := range w.senders {
+		wg.Add(1)
+		go func(s *sender) {
+			defer wg.Done()
+			s.waitForInflight()
+		}(s)
+	}
+	wg.Wait()
```
This logic is out of place here. It feels like it should be in the select case of `flushChan`. Then the sending will be paired with the waiting.

The select statement will be uglier and bigger but that's fine.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +109,29 @@ func (w *StatsWriter) Run() {
+		wg.Add(1)
+		go func(s *sender) {
+			defer wg.Done()
+			s.waitForInflight()
```
Please make this made exported (since it's obviously not private anymore) and document it. `WaitForInflight` is a good name.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -49,32 +50,39 @@ type TraceWriter struct {
 // NewTraceWriter returns a new TraceWriter. It is created for the given agent configuration and
 // will accept incoming spans via the in channel.
 func NewTraceWriter(cfg *config.AgentConfig) *TraceWriter {
+
+	in := make(chan *SampledSpans, 1000)
+
```
Why did you move this out? Can we put it back? I don't see the var used anywhere.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +145,49 @@ func (w *TraceWriter) Run() {
+			w.drainAndFlush()
+			// Wait for encoding/compression to complete on each payload,
+			// and submission to senders
+			w.wg.Wait()
```
Let's move this into `drainAndFlush` it is needed in all cases. And remove it from `Stop()` method.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +145,49 @@ func (w *TraceWriter) Run() {
+	w.flushChan <- notify
+	<-notify
+
+	var wg sync.WaitGroup
+	for _, s := range w.senders {
+		wg.Add(1)
+		go func(s *sender) {
+			defer wg.Done()
+			s.waitForInflight()
+		}(s)
+	}
+	wg.Wait()
```
This here is also out of place. I feel like it's a duplication. Let's factor it out into a global function and call it inside the `select` case just like I proposed for the stats writer. Maybe something like `waitForSenders`

DarcyRaynerDD committed:
[serverless] implement more PR feedback

DarcyRaynerDD committed:
[serverless] more PR feedback

DarcyRaynerDD committed:
[serverless] fix comment

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -36,24 +37,32 @@ const (
+		stats:     &info.StatsWriterInfo{},
+		stop:      make(chan struct{}),
+		flushChan: make(chan chan struct{}),
+		// syncMode reports whether the writer should flush on its own or only when FlushSync is called
```
Why is the documentation duplicated here? Should we remove it?

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -49,32 +50,37 @@ type TraceWriter struct {
 	traces       []*pb.APITrace // traces buffered
 	events       []*pb.Span     // events buffered
 	bufferedSize int            // estimated buffer size
+	syncMode     bool           // when enabled, traces are only sent when Flush is called
```
```suggestion
	syncMode     bool           // when enabled, traces are only sent when FlushSync is called
```
Would you mind creating a separate group of fields (just like we did in the stats writer) with the fields that are used only in sync mode? I mean `syncMode` and `flushChan`

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -49,32 +50,37 @@ type TraceWriter struct {
 // NewTraceWriter returns a new TraceWriter. It is created for the given agent configuration and
 // will accept incoming spans via the in channel.
 func NewTraceWriter(cfg *config.AgentConfig) *TraceWriter {
+
```
You forgot this line.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -139,6 +143,37 @@ func (w *TraceWriter) Run() {
+			w.addSpans(pkg)
+		case notify := <-w.flushChan:
+			w.drainAndFlush()
+			WaitForSenders(w.senders)
```
This is a lot better in my opinion. It still feels a little bit out of place however. It's like a sudden switch of context between two things. I wish this code was closer to where we actually use `w.senders`, but that can get really complex. Regardless, I think we've come a long way and I hope you're also happy with the outcome.

No action to take here, just rambling and being open to any ideas/suggestions.

DarcyRaynerDD replied:
We could potentially add an argument to drainAndFlush, eg `func drainAndFlush(waitForSenders bool)`? It's still one step separated from the flush call where we interact with the senders.

gbbr replied:
Your idea lead me to another (point being to have all the senders logic together). What if we changed `sendPayloads` to:
```go
func sendPayloads(senders []*sender, p *payload, sync bool) {
```
And then call it as `sendPayloads(w.senders, p, w.syncMode)` ?

Would there be any risks with this? I think the premise is to wait for each flush in sync mode anyway. If you see no problems with this and like it too, perhaps we should just absorb the `WaitForSenders` logic into `sendPayloads` then.

DarcyRaynerDD replied:
I've added a syncMode parameter to the sendPayloads function, and moved the sync logic in there.

gbbr commented on pkg/trace/writer/trace.go:
```diff
@@ -160,6 +195,22 @@ func (w *TraceWriter) addSpans(pkg *SampledSpans) {
+	w.flush()
+	// Wait for encoding/compression to complete on each payload,
+	// and submission to senders
+	w.wg.Wait()
```
I think you forgot to remove this line from the `Stop` method, now that we're doing it here, it shouldn't be needed there. Right?

gbbr commented on pkg/trace/writer/stats_test.go:
```diff
@@ -183,6 +183,71 @@ func TestStatsWriter(t *testing.T) {
+		}
+		assertPayload(assert, expectedHeaders, testSets, srv.Payloads())
+	})
+	t.Run("stop", func(t *testing.T) {
```
Please add an empty line above this.

gbbr commented on pkg/trace/writer/stats.go:
```diff
@@ -95,6 +111,19 @@ func (w *StatsWriter) Run() {
 	}
 }
 
+// FlushSync is a no-op for TraceWriter
```
Can we update this one too please?

gbbr reviewed (commented):
I think we're close to completion 🙂

DarcyRaynerDD committed:
[serverless] more PR feedback

gbbr reviewed (commented):
This is looking really good Darcy. Thanks for going with me through all the iterations. Left another potential improvement idea to consider.

DarcyRaynerDD committed:
[serverless] move sync logic into sender

DarcyRaynerDD committed:
Merge branch 'master' into darcy.rayner/trace-lambda-extension

DarcyRaynerDD committed:
[serverless] fix sendPayloads syncMode logic

gbbr commented on pkg/trace/writer/trace_test.go:
```diff
@@ -158,3 +158,90 @@ func payloadsContain(t *testing.T, payloads []*payload, sampledSpans []*SampledS
+}
+
+func TestTraceWriterSyncStop(t *testing.T) {
+	// TestTraceWriter
```
Let's remove this

gbbr commented on pkg/trace/writer/trace_test.go:
```diff
@@ -158,3 +158,90 @@ func payloadsContain(t *testing.T, payloads []*payload, sampledSpans []*SampledS
 }
+
+func TestTraceWriterFlushSync(t *testing.T) {
+	// TestTraceWriter
```
And this

gbbr reviewed (commented):
As far as I'm concerned this PR is good to go. Just a few nits.

DarcyRaynerDD committed:
[serverless] pr feedback

remeh commented on pkg/serverless/serverless.go:
```diff
@@ -223,7 +224,7 @@ func ReportInitError(id ID, errorEnum ErrorEnum) error {
 // Note that for now, we only subscribe to INVOKE and SHUTDOWN messages.
 // Write into stopCh to stop the main thread of the running program.
-func WaitForNextInvocation(stopCh chan struct{}, statsdServer *dogstatsd.Server, id ID) error {
+func WaitForNextInvocation(stopCh chan struct{}, statsdServer *dogstatsd.Server, traceAgent *traceAgent.Agent, id ID) error {
```
nit: I'm wondering if we should not have the `*Daemon` here instead of the `statsdServer` and `traceAgent`, wdyt?

DarcyRaynerDD replied:
I think it makes sense to refactor the daemon a bit so that it has a shutdown method that deals with the statsDServer and traceAgent. We can probably do that in a follow up PR though.

remeh reviewed (approved):
Just a last small nit but LGTM!
